### ISSUE TITLE
Automatically connecting plugs based on names

### ISSUE THREAD

**Issue by neuneck** [PaulSchweizer/flowpipe#56]
I have an idea for an API simplification. I know little about other people's use case, but for me, it makes a lot of sense to name plugs based of whatever semantic information flows through them ("features", "user_image", "detected_objects").

Since I found myself naming output and input plugs that I was going to connect the same way, I think it makes sense to extend the use of the `connect` method, as well as the `>>` operator to nodes, mapping output plugs of the node on the left to input plugs of the node on the right based on name equality.

The connections from the example would then simplify to
``` python
current_time >> van  # Note that the current_time.output and van.input plugs have the same name
current_time >> ldn
current_time >> muc
van.outputs['converted_time'] >> world_clock.inputs['time1']
ldn.outputs['converted_time'] >> world_clock.inputs['time2']
muc.outputs['converted_time'] >> world_clock.inputs['time3']
```
One could even go a step further an allow a similar syntax on single input/output situations, since the intended connection is unique, too, which would further simplify the API to
``` python
current_time >> van
current_time >> ldn
current_time >> muc
van >> world_clock.inputs['time1']  # Note that van only has one output plug
ldn >> world_clock.inputs['time2']
muc >> world_clock.inputs['time3']
```

If such a node-level connection shortcut is used but no connection was created, I would issue a warning, as this situation is probably not the intention of the user.

**Comment by PaulSchweizer** [comment]
I like this. 
I think I would impose a limit to the node >> node connection in a sense that it only connects anything if all outputs match all inputs so we don't end up with situations where only some connections have been established but some have not. But you might have planned it that way already anyways.

I also think that this could only apply to normal plugs, not to compound plugs as they are a bit unpredictable, haven't thought it through 100% though, so you might find a way to do it, but I think we can safely exclude them from this feature for now as it's not crucial and they are a special use case anyways.

**Comment by neuneck** [comment]
Thank you for your positive feedback!

For my taste, the restriction to cases where all inputs match all outputs goes too far. My goal was never having to refer to `inputs` or `outputs` explicitly if I name my plugs properly. For situations where unique plug names are not possible or more fine-grained control is required, the explicit output/input API would stay in place, of course.

**Comment by PaulSchweizer** [comment]
I see what you mean, for you, plugs are also somewhat synonymous with "data types", makes total sense to then just connect the ones that match and just ignore those that don't. Please go ahead, my idea was just a consideration. Thanks again for the proposal

**Comment by neuneck** [comment]
I was finally starting to look at this today. I noticed that I only ever use the `>>` operator as
```
output >> input
```
which makes sense, since we're building a directed acyclic graph, and the `>>` operator indicates the flow of information through the graph. Looking at the connection code, I was surprised that
```
input >> output
```
gives the same result, even though the flow direction is the other way.

I understand this is a design decision to reserve `<<` for disconnections. Now I have never used a disconnect in my graphs, since I define them once and then don't modify them after setting them up. 

In my opinion, syntactic sugar is great as long as the behavior is still explicit and intuitive. 
I think the connection sugar from output to input using the `>>` operator is amazing - it is brief and a perfect visual representation of the data _flow_ that's being set up. That the same operator with the plugs swapping places does the same thing is confusing to me, as the operator shape implies it's asymmetric. Having `input << output` connect the two might be better, however that would be a breaking change w.r.t. current behavior where `<<` disconnects two plugs. 
Regarding the disconnection operator `<<`, I understand the intention of providing an inverse operator to `>>`. Nevertheless, I think it's not a great visual representation of what's going on (it is stopping the flow of data, not reversing it), and I think the explicit `disconnect()` method should be preferred..

My suggestion would be to deprecate all uses of the `<<` operator and the use of the `>>` operator on `InputPlug` and `SubInputPlug`. I would leave `connect(self, other)` and `disconnect(self, other)` interfaces in place, of course, since they are great and explicit.

If only `OutputPlug` (and `SubOutputPlug`) had the `>>` sugar, it would directly represent a directed edge in the graph. Moreover, following this thread's suggestion of automatic connections based on plug names would add the `>>` operator to nodes, where it would mean "this node is upstream of that Node or InputPlug".

Of course, this is just a suggestion and I'm aware it's a bold one. I do think it's a change that would benefit flowpipe's API, though. I'm very much interested to hear what others have to say about this.

**Comment by PaulSchweizer** [comment]
I have also never used the disconnect feature myself. The main use case for it would be mostly in a visual editor where users create graphs interactively. So the << feature can and should be deprecated. 
I think I added that reversed output >> input feature for ease of use, but frankly, I have never used it that way and your concerns are all valid. So yes, let's deprecate that wrong direction flow as well.
Thanks for this suggestion, it does make things more clear and removes unnecessary features, I like it! Flowpipe wants to be simple and lightweight, so any unnecessary baggage we can drop is a success

**Comment by neuneck** [comment]
Closed by PR #138
